lib/services/access_request.go

```diff
@@ -245,53 +344,212 @@ func NewAccessRequest(user string, roles ...string) (AccessRequest, error) {
+			return trace.Wrap(err)
+		}
+		m.Roles.Allow = append(m.Roles.Allow, ma)
+	}
+
+	mas, err := role.GetAccessRequestConditions(Allow).GetTraitMappings().TraitsToRoleMatchers(m.traits)
+	if err != nil {
+		return trace.Wrap(err)
 	}
-	roleStates := make(map[string]rstate, len(req.GetRoles()))
-	for _, r := range req.GetRoles() {
-		roleStates[r] = rstate{false, false}
+
+	m.Roles.Allow = append(m.Roles.Allow, mas...)
+
+	if m.state.IsPending() {
+		for ka, va := range role.GetAccessRequestConditions(Allow).Annotations {
+			var vals []string
+		Avals:
+			for _, a := range va {
+				applied, err := applyValueTraits(a, m.traits)
+				if err != nil {
+					continue Avals
+				}
+				vals = append(vals, applied...)
+			}
+			m.Annotations.Allow[ka] = append(m.Annotations.Allow[ka], vals...)
+		}
 	}
+	return nil
+}
+
+// CanRequestRole checks if a given role can be requested.
+func (m *requestValidator) CanRequestRole(name string) bool {
+	for _, deny := range m.Roles.Deny {
+		if deny.Match(name) {
+			return false
+		}
+	}
+	for _, allow := range m.Roles.Allow {
+		if allow.Match(name) {
+			return true
+		}
+	}
+	return false
+}
+
+// SystemAnnotations calculates the system annotations for a pending
+// access request.
+func (m *requestValidator) SystemAnnotations() map[string][]string {
+	annotations := make(map[string][]string)
+	for k, va := range m.Annotations.Allow {
+		var filtered []string
+		for _, v := range va {
+			if !utils.SliceContainsStr(m.Annotations.Deny[k], v) {
+				filtered = append(filtered, v)
+			}
+		}
+		if len(filtered) == 0 {
+			continue
+		}
+		annotations[k] = filtered
+	}
+	return annotations
+}
+
+// ValidateAccessRequest validates an access request against the associated users's
+// *statically assigned* roles. If expandRoles is true, it will also expand wildcard
+// requests, setting their role list to include all roles the user is allowed to request.
+// Expansion should be performed before an access request is initially placed in the backend.
+func ValidateAccessRequest(getter UserAndRoleGetter, req AccessRequest, expandRoles bool) error {
+	user, err := getter.GetUser(req.GetUser(), false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var requireReason bool
+	validator := newRequestValidator(user.GetTraits(), req.GetState())
+
 	for _, roleName := range user.GetRoles() {
 		role, err := getter.GetRole(roleName)
 		if err != nil {
 			return trace.Wrap(err)
 		}
-	Allow:
-		for _, r := range role.GetAccessRequestConditions(Allow).Roles {
-			s, ok := roleStates[r]
-			if !ok {
-				continue Allow
-			}
-			s.allowed = true
-			roleStates[r] = s
+		if err := validator.push(role); err != nil {
+			return trace.Wrap(err)
+		}
+		if role.GetOptions().RequestAccess.RequireReason() {
+			requireReason = true
+		}
+	}
+
+	if requireReason && req.GetRequestReason() == "" {
+		return trace.BadParameter("request reason must be specified")
+	}
+
+	if r := req.GetRoles(); len(r) == 1 && r[0] == "*" {
```

Comment:
This function needs comments around each block of logic. For example, explain what `len(r) == 1 && r[0] == "*"` is.